<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="William Hagedorn-Rasmussen">

<title>Exploring CD8+ T-cell Specificities using Single Cell Immune Profiling with an Outlook to Reproducible Bio Data Science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="TCRSequenceFunctions_files/libs/clipboard/clipboard.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/quarto.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/popper.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/anchor.min.js"></script>
<link href="TCRSequenceFunctions_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="TCRSequenceFunctions_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="TCRSequenceFunctions_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="TCRSequenceFunctions_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="TCRSequenceFunctions_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-Import" id="toc-sec-Import" class="nav-link active" data-scroll-target="#sec-Import"> <span class="header-section-number">1</span> Imports</a></li>
  <li><a href="#sec-Intro" id="toc-sec-Intro" class="nav-link" data-scroll-target="#sec-Intro"> <span class="header-section-number">2</span> Introduction</a></li>
  <li><a href="#sec-TidyData" id="toc-sec-TidyData" class="nav-link" data-scroll-target="#sec-TidyData"> <span class="header-section-number">3</span> Tidying the data</a>
  <ul class="collapse">
  <li><a href="#sec-CleaningData" id="toc-sec-CleaningData" class="nav-link" data-scroll-target="#sec-CleaningData"> <span class="header-section-number">3.1</span> Cleaning</a></li>
  <li><a href="#sec-AugmentingData" id="toc-sec-AugmentingData" class="nav-link" data-scroll-target="#sec-AugmentingData"> <span class="header-section-number">3.2</span> Augmenting</a></li>
  </ul></li>
  <li><a href="#sec-ModelData" id="toc-sec-ModelData" class="nav-link" data-scroll-target="#sec-ModelData"> <span class="header-section-number">4</span> Modelling</a></li>
  <li><a href="#sec-Shiny" id="toc-sec-Shiny" class="nav-link" data-scroll-target="#sec-Shiny"> <span class="header-section-number">5</span> Shiny Integration</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  <li><a href="#sec-Appendix" id="toc-sec-Appendix" class="nav-link" data-scroll-target="#sec-Appendix"> <span class="header-section-number">6</span> Appendix</a>
  <ul class="collapse">
  <li><a href="#sec-AppendixA" id="toc-sec-AppendixA" class="nav-link" data-scroll-target="#sec-AppendixA"> <span class="header-section-number">6.1</span> Appendix A</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploring CD8+ T-cell Specificities using Single Cell Immune Profiling with an Outlook to Reproducible Bio Data Science</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>William Hagedorn-Rasmussen </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<section id="sec-Import" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Imports</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TCRSequenceFunctions)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data_list <span class="ot">&lt;-</span> <span class="fu">list</span>(data_donor_one_raw, data_donor_two_raw,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                  data_donor_three_raw, data_donor_four_raw)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>rows_raw <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(data_list)) {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  rows_raw <span class="ot">&lt;-</span> <span class="fu">append</span>(rows_raw, <span class="fu">nrow</span>(data_list[[i]])) </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(rows_raw)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 46526
[1] 46526 77854
[1] 46526 77854 37824
[1] 46526 77854 37824 27308</code></pre>
</div>
</div>
</section>
<section id="sec-Intro" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Introduction</h1>
<p>This package, <a href="https://github.com/WilliamH-R/TCRSequenceFunctions">TCRSequenceFunctions</a>, is a collection of functions made for working with data sets from a Single Cell Immune Profiling experiment made by 10x Genomics <span class="citation" data-cites="10XGenomics2022">[<a href="#ref-10XGenomics2022" role="doc-biblioref">1</a>]</span>. There is a total of four data sets which all follow the same general structure. They differ in that, they contain data from each their own respective donor.</p>
<p>The data sets contains binding counts between the donors’ library of T-Cell Receptors (TCRs) and a set of peptide-major histocompatibility complexes (pMHCs). The before-mentioned binding counts are so called unique molecular identifier (UMI) counts. For an explanation of all columns see <a href="#sec-AppendixA">Section&nbsp;6.1</a>.</p>
<p>In the following sections, each of the functions contained in <a href="https://github.com/WilliamH-R/TCRSequenceFunctions">TCRSequenceFunctions</a> will be explained, demonstrated and reasoned for. Generally, they can be divided into three types: Cleaning, Augmenting and Modelling where the main goal of the two first is to make the data tidy.</p>
<p>Lastly will be a short section on a Shiny Package, <a href="https://github.com/WilliamH-R/TCRSequenceShiny">TCRSequenceShiny</a>, which utilizes these functions to make a user-friendly interactive interface for data exploration.</p>
</section>
<section id="sec-TidyData" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Tidying the data</h1>
<p>The aim of tidying the data is to enable the data handling, and to ensure a reproducible result. Firstly, the data is cleaned e.g.&nbsp;by making sure, all cells only contain one piece of information. Afterwards, some augmented was needed to enable the modelling. This was done by e.g.&nbsp;adding new columns. A wrapper function was used to run all the preparation functions: <code>run_all_prep()</code>. This wrapper simply takes one of the raw data files included in the package as input, and pipe it through all the preparation functions, and output tidy data as in <a href="#tbl-tidydata">Table&nbsp;3.1</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>tidy_data <span class="ot">&lt;-</span> data_donor_one_raw_mock <span class="sc">%&gt;%</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">run_all_prep</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">kable</span>(<span class="fu">head</span>(tidy_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="tbl-tidydata" class="anchored">
<table class="table table-sm table-striped">
<caption>Table 3.1: Tidy data created from raw file</caption>
<thead>
<tr class="header">
<th style="text-align: left;">barcode</th>
<th style="text-align: left;">TCR_sequence</th>
<th style="text-align: left;">TCR_combination</th>
<th style="text-align: left;">donor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">TTGCCGTTCCGAATGT-14</td>
<td style="text-align: left;">CADPSGSARQLTF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
<tr class="even">
<td style="text-align: left;">TTGCCGTTCCGAATGT-14</td>
<td style="text-align: left;">CADPSGSARQLTF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TTGCCGTTCCGAATGT-14</td>
<td style="text-align: left;">CASSQEAGAATGELFF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
<tr class="even">
<td style="text-align: left;">TTGCCGTTCCGAATGT-14</td>
<td style="text-align: left;">CASSQEAGAATGELFF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TTGCCGTTCCGCGCAA-8</td>
<td style="text-align: left;">CVGDGGSQGNLIF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
<tr class="even">
<td style="text-align: left;">TTGCCGTTCCGCGCAA-8</td>
<td style="text-align: left;">CASSEGGFHPLHF</td>
<td style="text-align: left;">one_alpha_one_beta</td>
<td style="text-align: left;">donor1</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<section id="sec-CleaningData" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="sec-CleaningData"><span class="header-section-number">3.1</span> Cleaning</h2>
<p>As mentioned above, cleaning the data is mostly focusing on handling already present data and/or re-structure the data frame. The list of cleaning functions are as follows:</p>
<ol type="1">
<li><code>remove_unnecessary_columns()</code></li>
<li><code>split_TCR_sequences_find_non_promiscuous()</code></li>
<li><code>pivot_longer_TCR_sequences()</code></li>
<li><code>add_chain_ident_remove_prefix()</code></li>
<li><code>pivot_longer_pMHC()</code></li>
<li><code>tidy_pMHC_names()</code></li>
</ol>
<p>The first function takes the raw data frame as input, and simply removes the unnecessary columns as these arent needed. By default, the columns removed are those containing <code>"_binder"</code> and the column <code>"cell_clono_cdr3_nt"</code>. The dimensions of the data sets are then changed as shown in <a href="#tbl-function_one_dim">Table&nbsp;3.2</a></p>
<div id="tbl-function_one_dim" class="anchored">
<table class="table">
<caption>Table 3.2: Dimensions of data sets before and after being cleaned by <code>remove_unnecessary_columns()</code></caption>
<colgroup>
<col style="width: 20%">
<col style="width: 21%">
<col style="width: 19%">
<col style="width: 19%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th>Name of data set</th>
<th>Rows before</th>
<th>Rows after</th>
<th>Columns before</th>
<th>Columns after</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>data_donor_one_raw</code></td>
<td>46526</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>data_donor_two_raw</code></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td><code>data_donor_three_raw</code></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td><code>data_donor_four_raw</code></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sec-AugmentingData" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="sec-AugmentingData"><span class="header-section-number">3.2</span> Augmenting</h2>
<ol type="1">
<li><code>add_max_non_specific_binder()</code></li>
<li><code>evaluate_binder()</code></li>
<li><code>add_TCR_combination_identifier()</code></li>
</ol>
</section>
</section>
<section id="sec-ModelData" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Modelling</h1>
</section>
<section id="sec-Shiny" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Shiny Integration</h1>
</section>
<section id="references" class="level1 unnumbered">
<h1 class="unnumbered">References</h1>
</section>
<section id="sec-Appendix" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Appendix</h1>
<section id="sec-AppendixA" class="level2 unnumbered" data-number="6.1">


</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">6.1 Appendix A</h2><div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-10XGenomics2022" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">[1] </div><div class="csl-right-inline">10X Genomics. 2022. <span class="nocase">A New Way of Exploring Immunity - Linking Highly Multiplexed Antigen Recognition to Immune Repertoire and Phenotype</span>. <em>10x Genomics</em> (2022), 1–13. Retrieved from <a href="https://www.10xgenomics.com/resources/document-library/a14cde">https://www.10xgenomics.com/resources/document-library/a14cde</a></div>
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>