<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="William Hagedorn-Rasmussen">

<title>Exploring CD8+ T-cell Specificities using Single Cell Immune Profiling with an Outlook to Reproducible Bio Data Science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="TCRSequenceFunctions_files/libs/clipboard/clipboard.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/quarto.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/popper.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="TCRSequenceFunctions_files/libs/quarto-html/anchor.min.js"></script>
<link href="TCRSequenceFunctions_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="TCRSequenceFunctions_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="TCRSequenceFunctions_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="TCRSequenceFunctions_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="TCRSequenceFunctions_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abstract" id="toc-abstract" class="nav-link active" data-scroll-target="#abstract"> <span class="header-section-number">1</span> Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction"> <span class="header-section-number">2</span> Introduction</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods"> <span class="header-section-number">3</span> Methods</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"> <span class="header-section-number">4</span> Results</a></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion"> <span class="header-section-number">5</span> Discussion</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"> <span class="header-section-number">6</span> Conclusion</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploring CD8+ T-cell Specificities using Single Cell Immune Profiling with an Outlook to Reproducible Bio Data Science</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>William Hagedorn-Rasmussen </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<section id="abstract" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Abstract</h1>
<p>This is supposed to be an abstract.</p>
</section>
<section id="introduction" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Introduction</h1>
<p>With all the advancements in recent years in regards to computational power, laboratory techniques and -equipment, the amount of produced data is ever increasing. The focus in this article will be on biological data, even though it applies to a range of fields. For a reference as to how fast the development is, the cost of sequencing the entire human genome decreased from $70M to $2000 in a matter of 15 years <span class="citation" data-cites="Lunshof2022">(<a href="#ref-Lunshof2022" role="doc-biblioref">Lunshof et al. 2022</a>)</span> (Lunshof, INSERT reference). As a more recent discovery, the field of Single Cell Transcriptomics is emerging with a high increase in number of associated publications (Angerer, INSERT reference).</p>
<p>When a bioinformatician works with data, it is important it is done in a reproducible fashion. Exactly as in the laboratory, it is paramount that other scientists is able to achieve the same results and to come to the same conclusions. That tends to be an issue, as the code used for working with data is not necessarily made available. If that’s the case, how can you ever be sure, the no mistakes were made, or that any other problem occurred? Thereby, it is not only important to be working reproducibly, but also transparently.</p>
<p>A part of working with data is to find meaningful correlations, make new discoveries or support known hyptheseses. These findings should be presentable as to better share said findings. Especially when presenting to non-data stakeholders this becomes important, as they won’t necessarily have the same foundation of knowledge as your peers.</p>
<p>We can further extent the issue of presenting to stakeholders. It is likely that they want to see certain parts of the data which wasn’t included in the presentation. Then, if the data is made interactive, the stakeholders is enabled to find this themselves. Even without the need of writing a single line of code. In the end, this saves time and effort for both parties.</p>
<p>All of the above-mentioned issues I wish to address and take into account when working with this project. This is e.g.&nbsp;done by working transparently. Every pieces of code from the data preparation to the modelling is gathered in a package which is made available on GitHub. The functions written also include thorough documentations to make understanding the code easier. Through this, the data analyses are fully reproducible and can be reused to try and handle the increasing amount of data. Finally, the results of the modelling is made available through a Shiny App to make the data handling interactive.</p>
<ul>
<li>What is the situation now / why is this important?
<ul>
<li><p>Difficult to reproduce data analyses</p></li>
<li><p>Amount of available data grows</p></li>
<li><p>The presentation of data to non-data stakeholders</p>
<ul>
<li><p>E.g. in industry</p></li>
<li><p>Make it interactive to reduce work load for bioinformatician and stakeholders</p></li>
</ul></li>
</ul></li>
<li>What can this project do to help the issue?</li>
<li>What kind of data am I working with?
<ul>
<li><p>Needed theory to understand</p>
<ul>
<li><p>How are T-cells formed</p></li>
<li><p>What determines the TCR-sequence and what is the TCR</p></li>
<li><p>Alpha- beta sequence description</p></li>
<li><p>MHC, pMHC</p></li>
<li><p>Single Cell Immune Profiling</p></li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="methods" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Methods</h1>
<ul>
<li><p>Description of the structure of the package?</p></li>
<li><p>Description of data set?</p></li>
<li><p>How I cleaned</p>
<ul>
<li>Describe dimensions of the data set(s)</li>
</ul></li>
<li><p>How I augmented</p>
<ul>
<li>Describe dimensions of the data set(s)</li>
</ul></li>
<li><p>The integration into Shiny?</p></li>
</ul>
</section>
<section id="results" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Results</h1>
<ul>
<li>Mention pointers from method
<ul>
<li>What can the package offer</li>
</ul></li>
<li>Presentation of models</li>
<li>How does the Shiny app work / what can it do</li>
</ul>
</section>
<section id="discussion" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Discussion</h1>
</section>
<section id="conclusion" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Conclusion</h1>
</section>
<section id="references" class="level1 unnumbered">


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Lunshof2022" class="csl-entry" role="doc-biblioentry">
Lunshof, Jeantine E., Jason Bobe, John Aach, Misha Angrist, Joseph V. Thakuria, Daniel B. Vorhaus, Margret R. Hoehe, and George M. Church. 2022. <span>“<span class="nocase">Personal genomes in progress: from the Human Genome Project to the Personal Genome Project</span>.”</span> <em>Https://Doi.org/10.31887/DCNS.2010.12.1/Jlunshof</em> 12 (1): 44–57. <a href="https://doi.org/10.31887/DCNS.2010.12.1/JLUNSHOF">https://doi.org/10.31887/DCNS.2010.12.1/JLUNSHOF</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>